hybrasyl/Messaging/Admin.cs

```diff
@@ -0,0 +1,410 @@
+                                script.CompilationError);
+                        if (script.LastRuntimeError != string.Empty)
+                            errorSummary = string.Format("{0} runtime error: {1}", errorSummary,
+                                script.LastRuntimeError);
+                    }
+
+                    // Report to the end user
+                    return Success($"{scriptStatus}\n\n{errorSummary}", MessageTypes.SLATE_WITH_SCROLLBAR);
+                }
+            }
+            return Fail("Script {args[1].Trim()}: not found.");
+        }
+    }
+
+    class RollcharCommand : ChatCommand
+    {
+        public new static string Command = "rollchar";
+        public new static string ArgumentText = "<string class> <byte level>";
+        public new static string HelpText = "Simulate a new character with the specified class and level, and show resulting HP/MP.";
+        public new static bool Privileged = true;
+
+        public new static ChatCommandResult Run(User user, params string[] args)
+        {
+            if (!Enum.IsDefined(typeof(Enums.Class), args[1]))
+                return Fail($"Invalid class {args[0]}");
+
+            if (!byte.TryParse(args[2], out byte level) || level < 1 || level > Constants.MAX_LEVEL)
+                return Fail("Invalid level: must be between 1 and {Constants.MAX_LEVEL}");
+
+            // Create a fake User, and level them up to the desired level
+
+            var testUser = new User();
+            testUser.Map = new Map();
+            testUser.Stats.BaseHp = 50;
+            testUser.Stats.BaseMp = 50;
+            testUser.Class = (Enums.Class)Enum.Parse(typeof(Enums.Class), args[1]);
+            testUser.Stats.Level = 1;
+
+            while (testUser.Stats.Level < level)
+            {
+                testUser.GiveExperience(testUser.ExpToLevel);
+            }
+
+            return Success($"{testUser.Class}, Level {testUser.Stats.Level}, Hp: {testUser.Stats.BaseHp}, Mp: {testUser.Stats.BaseMp}");
+        }
+    }
+
+    class TeleportCommand : ChatCommand
+    {
+        public new static string Command = "teleport";
+        public new static string ArgumentText = "<string playername> | <uint mapnumber> <byte x> <byte y>";
+        public new static string HelpText = "Teleport to the specified player, or the given map number and coordinates.";
+        public new static bool Privileged = false;
+
+        public new static ChatCommandResult Run(User user, params string[] args)
+        {
+
+            if (args.Length == 1)
+            {
+                if (!Game.World.WorldData.ContainsKey<User>(args[1]))
+                {
+                    return Fail("That player cannot be found");
+                }
+                else
+                {
+                    var target = Game.World.WorldData.Get<User>(args[1]);
+                    user.Teleport(target.MapId, target.X, target.Y);
```

Comment:
MapId is not populated. Did you mean to use target.Map.Id?

Reply:
this should be target.Location.MapId

Reply:
good catch, this code is very old and predates `Location`. Updated.

Reply:
Location data seems to be off from the user.X / Y data. I'm in the ExampleInn, and I'm getting different co-ordinates.

![image](https://user-images.githubusercontent.com/33741983/35199583-eebfbf40-feb4-11e7-8420-0c7651ed7b8d.png)

Reply:
Stay tuned, I am refactoring this to address the concerns above.

Reply:
should be fixed